### ISSUE TITLE
Staging - Disk space % used on Data Partition has gone above 90

### ISSUE THREAD

**Issue by pdcp1** [responsible-ai-collaborative/aiid#3000]
I created this issue to collect all my findings in one place and take action based on that.

The storage issue is because of the `config` database (it is not directly related to the `aiidprod` database size).

![image](https://github.com/user-attachments/assets/c74e36a1-d093-4115-a8c0-38a5e9847529)

This `config` database is an internal MongoDB database controlled by the system itself. It contains system information about the cluster, sharding, operations, etc.

[Config database documentation reference](https://www.mongodb.com/docs/manual/reference/config-database/)

## `Config` database collections on Staging
| Collection                         | Size (GB)     |
|------------------------------------|---------------|
| config.system.preimages            | 5.6           |
| config.system.sessions             | 0.000164      |
| config.transactions                | 0.000084      |
| config.system.indexBuilds          | 0.000070      |
| config.image_collection            | 0.000042      |
| config.external_validation_keys    | 0.000019      |
| config.tenantMigrationDonors       | 0.000019      |
| config.tenantMigrationRecipients   | 0.000019      |

### Update

**What is the `config.system.preimages` ?**
There is no clear documentation about this system collection (this collection is not queryable, so we can't see its content), but my guess is that this collection stores all the pre-images of the documents required for our triggers.

**Why is this collection increasing in size?**
Every time that we restore the Production database to the Staging database, the `onIncidentUpdate` and `onNewIncident` triggers are fired. They are fired once for each incident document (more than 700 times at the same time) and both trigger crashes on Staging.
Then we received this email notification:

<img width="596" alt="image" src="https://github.com/user-attachments/assets/71380b7c-6575-4e51-aed7-4c178cb11d13">

The `onIncidentUpdate` trigger uses the pre-image of the document to compare and check if a report is added/removed or a new entity is associated with the current incident.
I believe the pre-images are stored in the `config.system.preimages` collection. When the trigger crashes, these pre-images are not cleaned up.

## TL;DR
We have a collection of old documents that have accumulated over time on each DB restore, and we are unable to clear them out. 🤦

**Comment by pdcp1** [comment]
@cesarvarela @kepae These are my findings regarding the Staging DB storage issue. Let's plan to discuss how to proceed next week.

**Comment by kepae** [comment]
>Pre-images are written to the config.system.preimages collection.
 >The config.system.preimages collection may become large. To limit the collection size, you can set expireAfterSeconds time for the pre-images as shown earlier.
 
Are we able to temporarily set `expireAfterSeconds` to a small value and see if they get cleaned up?

**Comment by pdcp1** [comment]
I'm currently in contact with the MongoDB Atlas support team.
I asked them to assist us with cleaning up the 'config' database, and their initial response was not the best 😒:

_Unfortunately that is not something we can do from our end, You may find the following documentation helpful._


We kept talking, and now they seem willing to assist us, although they take almost a day to respond to each of my messages. Let's hope for the best. 🤞

**Comment by pdcp1** [comment]
### From MongoDB Support:

Thank you for that information. From the shared output, `config.system.preimages` collection seems to be using most of the disk space.

`system.preimages 268033 2971281174 6032535552 11085 0`


The explanation is that there must be [Change Streams](https://www.mongodb.com/docs/v6.2/changeStreams/#change-streams) used on collections where pre/post image retrieval has been enabled. What this feature enables server is to keep a document version before it gets modified. That document version would be preserved under the `config.system.preimages` collection and could be retrieved through a Change Stream event if the appropriate option (`fullDocumentBeforeChange`: required) is set.



The likely reason why this collection happens to get so large is default retention policy is to keep the pre-image until the corresponding oplog entry remains available. Thus, if the oplog churn rate is slow and/or if the oplog size is large, then there could potentially be a large volume of pre-images kept around.

There are two ways to approach this issue:

1. You can use the collMod command to disable the pre-image capture for the change streams-enabled collections where this feature is not needed (that is if the change stream consumer does not need the pre-images). E.g.:
`db.runCommand( { collMod: "<changeStreamCollection>", changeStreamPreAndPostImages: { enabled: false } } )`

2. We can alter the [expireAfterSeconds](https://www.mongodb.com/docs/manual/reference/method/db.collection.watch/#change-streams-with-document-pre--and-post-images) option value to be more aggressive to change the expiration policy(say to 100 seconds). This is something that we would need to do for you as the [setParameter](https://www.mongodb.com/docs/atlas/unsupported-commands/#administrative) command is restricted in Atlas. If opting for this option, please let us know your desired value for this option along with when this change needs to be done by providing date, time window and timezone details.

It is worth noting that changing the retention threshold may not result into an immediate space reclamation. This is due to the storage engine behaviour where it only releases disk space when the free blocks are happen to be at the end of the data file.

Richard

MongoDB

**Comment by pdcp1** [comment]
I asked him to set the `expireAfterSeconds` value to 120 seconds since we don't need to store these pre-images for a longer period, 2 minutes is enough.
I will share his update as soon as I receive a confirmation response.

**Comment by pdcp1** [comment]
Good news!

The MongoDB Support team finally applied the `expireAfterSeconds` value to 120 seconds change and the problem is gone!
The size of the 'config' database has decreased significantly. 📉 

<img width="218" alt="Screenshot 2024-08-20 at 7 19 35 PM" src="https://github.com/user-attachments/assets/79cfc689-e9de-4c88-8fe0-7eba0fb277ca">
<img width="1568" alt="Screenshot 2024-08-20 at 7 19 11 PM" src="https://github.com/user-attachments/assets/c8f8bc21-47f4-4d19-adda-507f60e13255">


We are good now. Closing the issue ✅
